### ISSUE TITLE
TLS1.3 Using PSA and NON-PSA mode at the same time

### ISSUE THREAD

**Issue by tmordeko** [Mbed-TLS/mbedtls#8567]
Is it possible to use PSA and non-PSA mode (for client authentication) interchangeably during run-time.
Needs to support both mode simultaneously with the same stack running version 3.3.0.

**Comment by gilles-peskine-arm** [comment]
Could you clarify what you mean by “PSA and non-PSA mode”? And by “client authentication”, I guess you mean in TLS? TLS Client authentication relies on having a certificate signed by a CA, and that's independent of which crypto API is used under the hood to perform the verification.

**Comment by tmordeko** [comment]
Hi
Sorry for the confusion, let me rephrase my question.
AFAIU today the use of PKA APIs is based on enabling (in compile time) the following flag: #define MBEDTLS_USE_PSA_CRYPTO, so I was wondering if it is possible to interchangeably switch during run-time between PSA mode , for which PSA APIs are enabled working with external HW/keys to normal mode (to which I referred as non PSA). The question applies to both TLS1.2 and 1.3 currently working with version 3.3.0.

**Comment by gilles-peskine-arm** [comment]
TLS 1.3 always uses PSA crypto, and TLS 1.2 uses either PSA or legacy crypto based on `MBEDTLS_USE_PSA_CRYPTO`. There's no way to make the TLS code select between crypto APIs at run time.

However, this is a matter of which crypto API the TLS code uses. Unless you have PSA drivers (which for the most part can only work through the PSA API), it always calls the same crypto implementation.

So I don't understand what problem you're trying to solve here. What are you doing in “normal mode” that doesn't work with `MBEDTLS_USE_PSA_CRYPTO`?

**Comment by gilles-peskine-arm** [comment]
There is no way to choose at runtime whether a cryptography call from TLS will go through the PSA API or the legacy API. I don't know why this would be desirable anyway. Since they give the same result, why does it matter?

It is possible to include both a driver and the pure software path for a given algorithm, if you set up a driver with [fallback](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.5.1/docs/proposed/psa-driver-interface.md#fallback). This is not fully supported or well-tested yet, however. The intent is to support drivers that only support part of the functionality. For example, you could have a driver for an AES-128 accelerator, that falls back to software because you want AES-256 to work as well (just without accelerator benefit). It's not clear to me whether this is what you want?

**Comment by gilles-peskine-arm** [comment]
You can write an opaque driver (= secure element driver) for your hardware engine (instead of an accelerator driver). The decision to use an opaque driver or not depends on the key's lifetime. Opaque drivers are intended for use when you don't have the key in cleartext, only a wrapped encrypted copy of the key or a handle to the key that's stored inside the secure element. But if you want, you can make an opaque driver where the wrapped copy of the key is actually cleartext. Then operations will be performed in software by default, but if you create a key whose lifetime indicates the driver, operations on that key will use the hardware engine.

Alternatively, you can write an accelerator driver that sometimes returns `PSA_ERROR_NOT_SUPPORTED`, and enable fallback. Then it's up to the driver to decide at runtime whether to use the driver engine or not. But I'm not sure if we support fallback properly at the moment. The driver interface is not fully implemented and our priority was supporting software-only builds (which of course have always worked) and driver-only builds (i.e. builds where all the software implementation can be removed to save code size).

**Comment by tmordeko** [comment]
- Indeed with the secure element the key is not available on the CPU running the TLS stack, only its handle (uint32).
- There is a driver for the secure element (== is it considered an opaque driver ?) though not exactly PSA compliant.
- We need to support both modes (with and without secure element).

At the beginning I was thinking to replicate mbedtls_ecdsa_info struct, replacing some of the function pointers in it, like sign hash, with the secure element one (some of the functions will remain the same as the originals, like verify since we don't plan to offload it to the secure element) and a new enum for it of type mbedtls_pk_type_t; but now I can see that MBEDTLS_PK_OPAQUE is probably doing the same, isn't it ?
Regrading the decision choice between opaque driver or not can you please elaborate on the key's lifetime option you mentioned in your previous reply.

**Comment by tmordeko** [comment]
But what is my driver is not PSA compliant, can I still use MBEDTLS_PK_OPAQUE AFAIU they are tightly coupled.

**Comment by gilles-peskine-arm** [comment]
What do you mean by “not PSA compliant”? There's no official compliance for drivers because they're not an official PSA specification. Your driver needs to implement the [PSA driver interface](https://github.com/Mbed-TLS/mbedtls/blob/mbedtls-3.5.1/docs/proposed/psa-driver-interface.md).

**Comment by gilles-peskine-arm** [comment]
If you're just interested in the TLS stack, your application doesn't need to worry about pk types. Just tell the TLS stack which key you want to use with `mbedtls_ssl_conf_psk_opaque` or `mbedtls_ssl_set_hs_psk_opaque`.

**Comment by gilles-peskine-arm** [comment]
I'm closing this issue because it isn't clear what problem we're trying to solve, and in any case there will not be another Mbed TLS feature release with a “non-PSA mode”.
